player/lua/console.lua

```diff
@@ -1987,6 +2012,48 @@ mp.register_event('log-message', function(e)
             terminal_styles[e.level])
 end)
 
+mp.register_event('shutdown', function ()
+    if opts.persist_history == false or history_is_dirty == false then
+        return
+    end
+
+    history = histories[default_id]
+    local history_to_save = table.concat(history, '\n') .. '\n'
+    local history_path = mp.command_native({'expand-path', opts.history_path})
+
+    -- Reread the history file to get lines that other mpv instances may have
+    -- added.
+    local history_file = io.open(history_path)
+    if history_file then
+        local history_map = {}
+        for _, command in pairs(history) do
+            history_map[command] = true
+        end
+
+        for command in history_file:lines() do
```

Comment:
I think my previous opinion still stands. I don't think we should read all the file content on each save and instead append to history similar as watch-history works.

Reply:
Command history is different from watch history because console already deduplicates history every time a command is run so it's natural to save it like that. Do you mean to do a file write every time a command is run? Won't that make the history grow much larger when rerunning the same commands?

Reply:
Maybe this was bad example. What I think it should only save new items, but without rereading whole file again, it brings no real value. If say both instance of mpv uses the same new command worse case we get duplicated entry in history, no big deal. In other cases it won't happen.

Also there is a problem with concurrent saving of the history file by multiple instance of mpv at the same time. File might get corrupted.

Reply:
I still don't get it. First since history is deduplicated when running each command we would have to keep a separate copy of history to keep track of which commands are new. Then how would rerun commands be moved to the end if appending? For me the biggest utility of this is to close and reopen mpv and quickly rerun the last command when testing something.

Reply:
> First since history is deduplicated when running each command we would have to keep a separate copy of history to keep track of which commands are new.

Yes, a snapshot of the state of history during load.

> Then how would rerun commands be moved to the end if appending?

Why you want rerun command at the end of the file? New commands goes at the end when appending.

Reply:
You want rerun commands at the end because they were the commands ran last and I even offered the use case. Every shell history does this and also console history while it is running.

Reply:
Then don't de-duplicate so aggressively. Or store it in heap, sorted by the timestamp with time of last command use.

The point here is that appending makes it both more performant and resilient to errors. Both read/write and even reduces concurrency issue. If you prefer always rewriting whole file there is no point in using simple structure as line based history and instead it should be more complex structure that resolves all the problems.

I don't think any shell by default is de-duplicating history files, at least mine doesn't, except consecutive entries. Which is also a feature, because if you go up in history with up arrow you want to got through each command as they were typed and not only unique ones.

Reply:
Changed to append all commands entered after startup though this increases the history size and decreases reading performance instead, but it seems to be what shells do, except zsh which rewrites when you enable the more complex options.

By the way, I tried installing fish, and it deduplicates the read history by default. zsh does it with `HIST_IGNORE_ALL_DUPS`, but zsh has terrible defaults in general, see https://fishshell.com/docs/current/design.html#configurability-is-the-root-of-all-evil

Same for case insensitive completion, fish enables it by default, and I didn't understand why we didn't do it for console only to duplicate the worse defaults of bash and zsh.